ably/realtime/realtimepresence.py

```diff
@@ -0,0 +1,781 @@
+
+    async def leave_client(self, client_id: str, data: Any = None) -> None:
+        """
+        Leave presence on behalf of another clientId (RTP15).
+
+        Args:
+            client_id: The clientId to leave
+            data: Optional data to send with the leave message
+
+        Raises:
+            AblyException: If channel state prevents leaving or clientId mismatch
+        """
+        return await self._leave_client(client_id, data)
+
+    async def _enter_or_update_client(
+        self,
+        id: str | None,
+        client_id: str | None,
+        data: Any,
+        action: int
+    ) -> None:
+        """
+        Internal method to handle enter/update operations.
+
+        Args:
+            id: Optional presence message id
+            client_id: Optional clientId (if None, uses connection's clientId)
+            data: Optional data payload
+            action: The presence action (ENTER or UPDATE)
+
+        Raises:
+            AblyException: If connection/channel state prevents operation or clientId mismatch
+        """
+        channel = self.channel
+
+        # Check connection state
+        if channel.ably.connection.state not in [
+            ConnectionState.CONNECTING,
+            ConnectionState.CONNECTED,
+            ConnectionState.DISCONNECTED
+        ]:
+            raise AblyException(
+                f'Unable to {PresenceAction._action_name(action).lower()} presence channel; '
+                f'connection state = {channel.ably.connection.state}',
+                90001, 400
+            )
+
+        action_name = PresenceAction._action_name(action).lower()
+
+        log.info(
+            f'RealtimePresence.{action_name}(): '
+            f'channel = {channel.name}, '
+            f'clientId = {client_id or "(implicit) " + str(_get_client_id(self))}'
+        )
+
+        # RTP15f: Check clientId mismatch (wildcard '*' is allowed to enter on behalf of any client)
+        my_client_id = _get_client_id(self)
+        if my_client_id and client_id and client_id != my_client_id and my_client_id != '*':
+            raise AblyException(
+                f'Unable to {action_name} presence channel with clientId {client_id} '
+                f'as it does not match the current clientId {my_client_id}',
+                40012, 400
+            )
+
+        # RTP8c: Use connection's clientId if not explicitly provided
+        effective_client_id = client_id if client_id is not None else my_client_id
+
+        # Create presence message
+        presence_msg = PresenceMessage(
+            id=id,
+            action=action,
+            client_id=effective_client_id,
+            data=data
+        )
+
+        # Convert to wire format
+        wire_msg = presence_msg.to_encoded(cipher=channel.cipher)
+
+        # RTP8d/RTP8g: Handle based on channel state
+        if channel.state == ChannelState.ATTACHED:
+            # Send immediately
+            return await self._send_presence([wire_msg])
```

Comment:
there is one not very well described in the spec use case, that we need to cover. Even if channel is attached, connection itself can be disconnected and if presence message will be queued on the connection manager level, we'll resend it immediately after `CONNECT` without wait for attach, that will result in fail

Reply:
unfortunately, it won't work, not sure how to simulate it properly in test, but it should be very easy to reproduce on the actual client:
- create realtime client
- attach channel
- wait until attached
- turn off wifi (in test we can potentially dispose transport)
- send presence message (we should be in ATTACHED state while we DISCONNECTED)
- turn on wifi 
- wait for presence message response (expected success, actual failed)

I suggest creating additional PR to fix this